README.md
# Christopher v1.0.0-alpha.5

[![Build Status](https://travis-ci.org/davidderus/christopher.svg?branch=master)](https://travis-ci.org/davidderus/christopher)
[![Go Report Card](https://goreportcard.com/badge/github.com/davidderus/christopher)](https://goreportcard.com/report/github.com/davidderus/christopher)
[![GoDoc](https://godoc.org/github.com/davidderus/christopher?status.svg)](https://godoc.org/github.com/davidderus/christopher)

## Description

Christopher is your everyday direct-download companion.

It automatically grabs new episodes from RSS feeds, debrids their URL if needed,
and send them to a downloader service.

It also offers an integrated webserver for remote URL submission and accepts
debrid and download instruction from the command line.

## Usage

```shell
# Starts a FeedWatcher
christopher feed-watcher
# shorter version: christopher fw

# Runs a webserver with a simple interface to debrid and download URIs
christopher webserver
# shorter version: christopher ws

# Debrids an URI
christopher debrid "http://rapidgator.net/file/HTGAWM.mkv"
# shorter version: christopher de "http://rapidgator.net/file/HTGAWM.mkv"

# Downloads an URI
christopher download "https://google.fr"
# shorter version: christopher do "https://google.fr"

# Debrids and downloads an URI
christopher debrid-download "http://rapidgator.net/file/HTGAWM.mkv"
# shorter version: christopher dedo "http://rapidgator.net/file/HTGAWM.mkv"

# Use a custom config file (default in ~/.config/christopher/config.toml)
christopher -c ~/my/custom/config.toml […]
```

## Configuration

Christopher looks for a toml configuration file at
`$HOME/.config/christopher/config.toml`.

```toml
# Download configuration (required)
# The downloader is an external service Christopher pushes links to.
[downloader]
  # Name of the service downloading URIs
  name = "aria2"

  [downloader.auth_infos]
    token = "my-good-token"
    rpc_url = "http://127.0.0.1:6800/jsonrpc"

# Debrider configuration (optional)
# The debrider converts links from specific services to a downloadable link.
# Each link sent to Christopher is first tested against each debriders
# to see if it can be debrided.
[debrider]
  name = "AllDebrid"

  [debrider.auth_infos]
    username = "valid-username"
    password = "valid-password"
    base_url = "https://alldebrid.com"

# FeedWatcher configuration (optional)
# The feedwatcher watch some feeds and send every new links
# to the debriders/downloader
[feedwatcher]
  # Defining a custom watch interval in minutes (default to 30 min.)
  watch_interval = 5

  # Adding one feed for the feedwatcher to look for
  [[feedwatcher.feeds]]
    title = "DirectDownload Feed"

    # URL to the provider feed
    url = "https://directdownload.tv"

    provider = "DirectDownload"

# Providers configuration (optional)
# For each feed provider, you can setup some specific config like a list
# of host to send to the debrider.
[providers]
  # Provider config key and feed provider key must be the same
  [providers.DirectDownload]
    # If specified, favorite_hosts will be looked for in the provider links.
    # If none of them are found, nothing will be downloaded.
    # If not specified, the first link available is downloaded.
    favorite_hosts = ["uploaded.net", "rapidgator.net"]

# WebServer configuration (optional)
# The webserver accepts valid URLs and sent them to Christopher
# debriders/downloader
[webserver]
  # Setting a custom host (default to "127.0.0.1")
  host = "0.0.0.0"

  # Setting a custom webserver port (default to 8000)
  port = 8080

  # Defining a secret for the CSRF token generation (required)
  secret = "my-strong-secret"

  # Constraining CSRF cookie to be HTTPS only if true (default to false)
  secure_cookie = true

  # Setting a realm for the HTTP digest auth (default to "christopher.local")
  auth_realm = "download-helper.local"

  # Users are a list of allowed users.
	# If no users are given, no Digest auth is setup.
  [[webserver.users]]
  name = "johndoe"

  # Setting a password via a MD5 string using:
  # `echo -n "$username:download-helper.local:$password" | md5`
  # Also works with all the algorithms supported by HTTP Digest
  password = "36f0730e47562fbdf9b91434130f91f2"

# Teller configuration (optional)
# The Teller handles logging across the whole application.
# It supports text and JSON logging with variables.
[teller]
  # A level from when the Teller must log things
  # Default is `info`, meaning info logs and above will be shown.
  log_level = "debug"

  # The log items format
  # Default is `text`
  log_formatter = "json"
```

## Supported services

Here is a complete list of all supported services.

Do not hesitate to write a PR with some tests to add more.

### Providers

- DirectDownload (`provider = "DirectDownload" # or dd, directdownload, directdownload.tv`)

### Debriders

- AllDebrid (`name = "AllDebrid" # or alldebrid, Alldebrid, ad`)

### Downloaders

- Aria2 (`name = "aria2" # or Aria2, aria`)

## Upcoming features

- [x] A complete logger with log levels handling
- [ ] A download history to avoid duplicates and show status in webserver
- [ ] A successful download notifier (_push or email_)
- [ ] A lighter Docker Image
- [ ] A better communication between the webserver and Christopher core
- [ ] A documentation about the dispatcher and its stories/scenarios

## Docker

You can run Christopher with Docker with the following command:

```shell
# First build the image
docker build -t christopher .

# As a command line debrider
docker run -v $HOME/.config/christopher/config.toml:/christopher/config.toml:ro christopher debrid "http://rapidgator.net/file/HTGAWM.mkv"
```

In order to download files, you must give the christopher container access
to a Downloader.

Looking at the `docker-compose.yml` file, you will see a basic example of a
working christopher + aria2 setup. Try it with `docker-compose up`.

Remember that you need to update your christopher config file with the right
aria2 RPC token and use `http://aria2:6800/jsonrpc` as the `rpc_url`.

## Development

To build bindata file, use `go-bindata -pkg webserver -o webserver/bindata.go -prefix webserver webserver/templates/`.

## Licence

MIT Licence. Click [here](LICENCE) to see the full text.


CHANGELOG.md
# Changelog

## v1.0.0-alpha.5 - 2018/01/21

- Not using Travis for cross-platform deployment
- Fixing invalid yml indent in goreleaser config
- Using goreleaser default script and not relying on go for deploy

## v1.0.0-alpha.4 - 2018/01/21

- Adding execute bit to deploy script
- Removing osx from travis env config (the tests were always failing)

## v1.0.0-alpha.3 - 2018/01/21

- Fixing wrong shasum error for bootstrap css and js

## v1.0.0-alpha.2 - 2017/10/22

- Renaming config key `rpcURL` to `rpc_url` for consistency
- Introducing Teller, a text/json logger
- Releasing as binary for next releases
- Fixing licence copyright

## v1.0.0-alpha.1 - 2017/09/30

- Adding GoRelease config
- Adding Travis config
- Initial release


config/config.go
package config

import (
	"errors"
	"fmt"
	"os"
	"os/user"
	"path"

	"github.com/BurntSushi/toml"
)

const (
	// defaultWatchInterval defines the time in minutes between two FeedWatcher scans
	defaultWatchInterval = 30

	// defaultDBName is the default filename for the database
	defaultDBName = "database.db"

	// defaultHost sets up the webserver host default as a local host
	defaultHost = "127.0.0.1"

	// defaultPort sets up the webserver port to a non-standard one
	defaultPort = 8000

	// defaultAuthRealm is the default realm used for the web server digest authentication
	defaultAuthRealm = "christopher.local"

	// defaultLogLevel sets the minimum level for logging infos
	defaultLogLevel = "info"

	// defaultLogFormatter sets the log items formatter
	defaultLogFormatter = "text"
)

// Feed is a Feed Representation
type feed struct {
	Title    string // Remote feed title
	URL      string // URL to the feed
	Provider string // The feed provider
}

// FeedWatcherOptions defines some options for the FeedWatcher
type feedWatcherOptions struct {
	WatchInterval int `toml:"watch_interval"` // In Minutes
	Feeds         []*feed
}

// DownloaderOptions defines options for the downloader
type DownloaderOptions struct {
	Name            string
	AuthInfos       map[string]interface{} `toml:"auth_infos"`
	DownloadOptions map[string]interface{} `toml:"download_options"`
}

// DebriderOptions defines name and auth info for the debrider
type DebriderOptions struct {
	Name      string
	AuthInfos map[string]string `toml:"auth_infos"`
}

// ProviderOptions specify options for a given provider
type ProviderOptions struct {
	FavoriteHosts []string `toml:"favorite_hosts"`
}

type webUser struct {
	Name     string
	Password string
}

// WebServerOptions defines some of the webserver options
type WebServerOptions struct {
	Port int
	Host string

	// Secret is the secret for the CSRF token generation
	Secret string

	// SecureCookie constrains CSRF cookie to be HTTPS only if true
	SecureCookie bool `toml:"secure_cookie"`

	// AuthRealm is the realm for the HTTP digest auth
	AuthRealm string `toml:"auth_realm"`

	// Users are a list of allowed users.
	// If no users are given, no Digest auth is setup.
	Users []webUser
}

// TellerOptions defines logging options for the Teller
type TellerOptions struct {
	// LogLevel is a level from when the Teller must log things
	LogLevel string `toml:"log_level"`

	// LogFormatter is the log items format
	LogFormatter string `toml:"log_formatter"`
}

// Config defines the Christopher configuration
type Config struct {
	configPath string `toml:"config_path"`

	DBPath      string `toml:"db_path"`
	FeedWatcher feedWatcherOptions

	Downloader DownloaderOptions

	Debrider DebriderOptions

	Providers map[string]ProviderOptions

	WebServer WebServerOptions

	Teller TellerOptions
}

// Load loads config from the default config path
func Load() (*Config, error) {
	return LoadFromFile(DefaultConfigPath())
}

// LoadFromFile loads the configuration from a file
func LoadFromFile(configPath string) (*Config, error) {
	// Initializing configuration
	var config Config

	// Setting defaults for required elements
	config.setDefaults()

	// Parsing config file, thus overriding defaults if needed
	_, configError := toml.DecodeFile(configPath, &config)
	if configError != nil {
		if os.IsNotExist(configError) {
			return nil, fmt.Errorf("Missing config file in %s", configPath)
		}

		return nil, fmt.Errorf("Invalid config file: %v", configError)
	}

	// Validating resulting configuration
	validationError := config.validate()
	if validationError != nil {
		return nil, validationError
	}

	return &config, nil
}

// UserDir returns the user home directory
func UserDir() (string, error) {
	user, userError := user.Current()
	if userError != nil {
		return "", userError
	}

	return user.HomeDir, nil
}

// DefaultConfigPath returns the default path to the config file
func DefaultConfigPath() string {
	userDir, _ := UserDir()
	return path.Join(userDir, ".config", "christopher", "config.toml")
}

func (c *Config) validate() error {
	// Validating DBPath
	if c.DBPath == "" {
		return errors.New("DBPath can't be blank")
	}

	// Must have 32 bytes secret for CSRF protection
	if c.WebServer.Secret == "" {
		return errors.New("A 32 bytes secret token must be set")
	}

	return nil
}

func (c *Config) setDefaults() {
	userDir, _ := UserDir()

	// Setting default DBPath
	c.DBPath = path.Join(userDir, ".config", "christopher", defaultDBName)

	// Setting default WatchInterval
	c.FeedWatcher.WatchInterval = defaultWatchInterval

	// WebServer defaults
	c.WebServer.Host = defaultHost
	c.WebServer.Port = defaultPort
	c.WebServer.AuthRealm = defaultAuthRealm

	c.Teller.LogLevel = defaultLogLevel
	c.Teller.LogFormatter = defaultLogFormatter
}
